Adding keyboard shortcuts to common actions

Adaverse opened the pull request:
closes: #22870 

Hello,

Added below shortcuts to the DAGS grid screen - 
- At DagRun
  - Clear existing tasks (`shift + c`)
  - Mark as success or fail (`shift + s` or `shift + f`)
  - Toggle note (`shift + n`)
  - Edit notes (`shift + e`)
- At TaskInstance
  - Clear task (`shift + c`)
  - Mark as success or fail `(shift + s` or `shift + f`)
  - Toggle notes (`shift + n`)
  - Edit notes (`shift + e`)
- Toggle shortcut cheat sheet (`shift + /`)
 
Added the shortcut cheat sheet modal - 

![image](https://github.com/apache/airflow/assets/35839624/cdaecb71-f553-4078-a50a-0c4dc0ffb778)

The tasks 
- [x] Implementing various shortcuts for grid view
- [x] Creating a cheat sheet for the information
- [x] keeping focus on the action buttons of the opened modal by default

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Adaverse committed:
useKeysPress

Adaverse committed:
useKeysPress

bbovenzi commented on airflow/www/static/js/dag/details/taskInstance/taskActions/ClearInstance.tsx:
```diff
@@ -77,6 +78,10 @@ const ClearInstance = ({
 
+  const initialClearButtonFocusRef = useRef<HTMLButtonElement>(null);
+
+  useKeysPress("shiftKey", ['c', 'C'], onOpen);
```
Cool! Do you have a list of what shortcuts you want to do? We'll need to check this across browsers or toggle on/off if it conflicts with a user's shortcuts for something else.

Adaverse replied:
Hi @bbovenzi,

Added below shortcuts to the DAGS grid screen - 
- At DagRun
  - Clear existing tasks (`shift + c`)
  - Mark as success or fail (`shift + s` or `shift + f`)
  - Add note (`shift + n`)
- At TaskInstance
  - Clear task (`shift + c`)
  - Mark as success or fail `(shift + s` or `shift + f`)
  - Add note (`shift + n`)

The shortcuts are the same, just depends on what is selected. Will add more in the process.

Yes, I have checked with major browsers (Chrome, Firefox, Edge, Opera), and these shortcuts aren't used as of yet (At least nothing happened when I clicked :)).

Adaverse replied:
Hi @bbovenzi,
I have fixed all the issues. Can you have a look once at the implementation?

bbovenzi replied:
Looking good!
Do we want to add a cheatsheet somewhere?
Also, for adding a note, we should make sure to toggle the accordion open if it is closed.

Adaverse replied:
Thanks for having a look. 

Yes, working on creating the cheat sheet.

Yes, will make sure to toggle the notes accordion open (if closed).

Adaverse replied:
> Looking good! Do we want to add a cheatsheet somewhere? Also, for adding a note, we should make sure to toggle the accordion open if it is closed.

Hi @bbovenzi 
- Added a cheat sheet for shortcuts
- Yes, toggling the accordion and making the text area in focus

Adaverse committed:
Added various keyboard shortcuts in DAGS (grid) screen. Below is the list
- At DagRun
  - Clear existing tasks (shift + c)
  - Mark as success or fail (shift + s or shift + f)
  - Add note (shift + n)
- At TaskInstance
  - Clear task (shift + c)
  - Mark as success or fail (shift + s or shift + f)
  - Add note (shift + n)

Adaverse committed:
Added various keyboard shortcuts in DAGS (grid) screen. Below is the list
- At DagRun
  - Clear existing tasks (shift + c)
  - Mark as success or fail (shift + s or shift + f)
  - Add note (shift + n)
- At TaskInstance
  - Clear task (shift + c)
  - Mark as success or fail (shift + s or shift + f)
  - Add note (shift + n)

Adaverse committed:
static checks resolved

Adaverse committed:
static checks resolved

Adaverse committed:
Merge branch 'main' into test

Adaverse committed:
- added shortcut for edit notes, toggle notes accordion, toggle shortcut cheat sheet
- added cheat sheet modal
- using local storage variable to ignore event which text area is in focus

Adaverse committed:
- added shortcut for edit notes, toggle notes accordion, toggle shortcut cheat sheet
- added cheat sheet modal
- using local storage variable to ignore event which text area is in focus

Adaverse committed:
fixed static checks

Adaverse committed:
fixed static checks

Adaverse committed:
fixed notes accordion toggle logic

Adaverse committed:
fixed notes accordion toggle logic

Adaverse committed:
fixed test

Adaverse committed:
fixed test

Adaverse committed:
static checks fixed

Adaverse committed:
static checks fixed

bbovenzi commented on airflow/www/static/js/App.tsx:
```diff
@@ -55,6 +56,8 @@ interface AppProps extends PropsWithChildren {
   cache: EmotionCache;
 }
 
+const isInputInFocus: IsInputInFocus = "isInputInFocus";
```
it's just a string we don't need a special type for it.

```suggestion
const isInputInFocus = "isInputInFocus";
```

Adaverse replied:
Here the idea behind having a type for a string was to enforce this to have only specific values (since it is the key of the local storage object). In the above case, it was below.
```
// In type/index.ts
type IsInputInFocus = "isInputInFocus";
```
We can replace it with a global constant that will be imported wherever used. Where can I define a constant to import it anywhere required? I do not seem to find where we define all constants used in UI.

bbovenzi replied:
Just exporting the string from the useKeysPress is fine.

Adaverse replied:
Done. Defined in `useKeysPress` and importing it everywhere.

bbovenzi commented on airflow/www/static/js/dag/details/NotesAccordion.tsx:
```diff
@@ -106,8 +160,16 @@ const NotesAccordion = ({
+                    }}
                     data-testid="notes-input"
+                    onFocus={() => {
+                      localStorage.setItem(isInputIFocus, "true");
```
We shouldn't use localstorage for this. What are you trying to do here?

Adaverse replied:
As mentioned [here](https://github.com/apache/airflow/pull/30950#issuecomment-1553315956), no matter what I use, the events still end up bubbling when the user is typing in the notes text box. So, for example, if the user, while typing in the notes textbox presses `shift + C` ( to get capital C ), it fires up the corresponding shortcut as well. Hence, in order to avoid triggering the shortcut, we are using this LocalStorage to identify whether the user is typing or not at any given moment [in this line](https://github.com/apache/airflow/blob/275d4a226a3059bd50b5ae6417a2d4f506c5ae1f/airflow/www/static/js/utils/useKeysPress.ts#L43). If yes, then we will ignore the event to trigger a shortcut, or if no, then otherwise. And onBlur, we are setting it to false.

bbovenzi replied:
What if we just got rid of it as an accordion? The text has a limit so it will never be too long.
I can help with that part if you'd like

Adaverse replied:
It is not the problem with the accordion but with how React handles the events and here we have a mix of flask app builder with React inserted into it, which makes the situation even more complex. Even if we got rid of the accordion, given any text box user is typing (in any part of the whole app given useKeysPress components are mounted), the events will be bubbled to the `useKeysPress`, and it will perform those actions that will not be desirable to the user (only while typing). Hence, have put this check using local storage that will completely stop it.

bbovenzi replied:
Ok, we can keep it for now then.

bbovenzi commented on airflow/www/static/js/dag/Main.tsx:
```diff
@@ -145,6 +153,12 @@ const Main = () => {
   };
 
+  useKeysPress(
+    keyboardShortcutIdentifier.toggleShortcutCheatSheet.primaryKey,
```
Could we just pass `keyboardShortcutIdentifier.toggleShortcutCheatSheet` and `onToggleShortcut`?

Adaverse replied:
This is done

bbovenzi commented on airflow/www/static/js/dag/details/NotesAccordion.tsx:
```diff
@@ -80,12 +89,56 @@ const NotesAccordion = ({
+    keyboardShortcutIdentifier.viewNotes.secondaryKey,
+    () => {
+      // toggling notes accordion (index = 0)
+      toggleNotesAccordionIthIndex(0);
```
I don't think we need all this extra `toggleNotesAccordionIthIndex` just toggle the `editMode`

Adaverse replied:
True! But implemented `toggleNotesAccordionIthIndex`, thinking new panels might be included in the accordion in the future. Since you think that it might be overkill, hence replaced it with the simplest implementation - `toggleNotesPanel`

bbovenzi commented on airflow/www/static/js/dag/keyboardShortcutIdentifier.ts:
```diff
@@ -0,0 +1,71 @@
+  },
+};
+
+export default keyboardShortcutIdentifier;
```
Should we put a small prompt somewhere so users even know this exists?

Adaverse replied:
Added a button. Let me know your thoughts

![image](https://github.com/apache/airflow/assets/35839624/87540021-01d3-4424-acb6-9bdcefe0783b)

bbovenzi replied:
Text could be smaller, and we could just say : "Type `shift`+`/` for shortcuts

Adaverse replied:
Done.

![image](https://github.com/apache/airflow/assets/35839624/44a133eb-3229-4411-9ea6-20c688ac1c89)

bbovenzi reviewed (changes requested):
Left a few comments.

Adaverse committed:
Code reivew changes - reduced useKeysPress function signature

Adaverse committed:
Code review changes - reduced funtion signature of usekeysPress

Adaverse committed:
- removed toggleNotesAccordionIthIndex with more simple toggleNotesPanel
- fixed static check failure

Adaverse committed:
- Removed type defined a string type
- added isInputInFocus in useKeysPress and importing it from here everywhere

Adaverse committed:
added prompt for shortcut

bbovenzi reviewed (approved):
looking good!

Adaverse committed:
Merge branch 'main' into test
